### PR TITLE
[Model] RGCN mini-batch training; Better bipartite graph support

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

RGCN minibatch training for node classification task. See `entity_classify_mb.py`.

Better bipartite graph support:
* Upon a DGLHeteroGraph is created, it will check whether its metagraph is a uni-directional bipartite graph.
* If yes, then it will split the `ntypes` into `srctypes` and `dsttypes`. In this case, we allow two node types to have the same name as long as one belongs to `srctypes` and the other belongs to `dsttypes`.
* Following new APIs are added to ease the use of a bipartite graph:
  - `srctypes` and `dsttypes`
  - `srcdata` and `dstdata`
  - `srcnodes` and `dstnodes`
* Besides using the above APIs to distinguish nodes types in the SRC and DST groups, another way is by prepending the type name with `"SRC/"` or `"DST/"`. For example, `g.number_of_nodes("SRC/user")` returns the number of nodes of type `"user"` in the SRC group.

Other changes:
* Change the original `entity_classify.py` to directly use node embedding. Previous code use one embedding for one (srctype, etype) pair. Verified that the change does not influence accuracy.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by BarclayII** [comment]
One side question: should we consider launching multiple CUDA streams for `multi_update_all` (or even one stream for each relation)?  I saw PyG doing this in `HeteConv`.  Not very urgent (at least for this PR) though.

File: python/dgl/heterograph.py [hunk 1]
```diff
@@ -304,6 +335,24 @@ def add_edges(self, u, v, data=None, etype=None):
     # Metagraph query
     #################################################################
 
+    @property
+    def is_unibipartite(self):
+        """Return whether the graph is a uni-bipartite graph.
```
Comment (BarclayII) [inline python/dgl/heterograph.py]: We may want to reconsider how we call such graph structures.  At least "unidirectional bipartite graph" refers to the graphs with only one edge type and two node types in our docstrings.

**Comment by jermainewang** [comment]
> One side question: should we consider launching multiple CUDA streams for `multi_update_all` (or even one stream for each relation)? I saw PyG doing this in `HeteConv`. Not very urgent (at least for this PR) though.

From your benchmarks, the major bottleneck seems to be data movement between CPU and GPU, so we could put that as low priority.

File: python/dgl/heterograph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/heterograph.py]: I think uni-directional bipartite graph is the word for this kind. More formally, given the definition of uni-directional bipartite graph for homogeneous graph, we can extend this definition to heterogeneous graph by considering the metagraph -- a heterograph is uni-directional bipartite iff its metagraph is uni-directional bipartite.

A block graph is a special uni-bipartite graph with nodes properly relabeled and each DST node has one mirror in the SRC group.

File: python/dgl/heterograph.py [hunk 2]
```diff
@@ -479,16 +530,75 @@ def get_ntype_id(self, ntype):
...
+        """
+        if ntype is None:
+            if len(self._srctypes_invmap) != 1:
+                raise DGLError('SRC node type name must be specified if there are more than one '
+                               'SRC node types.')
+            return 0
+        ntid = self._srctypes_invmap.get(ntype, None)
+        if ntid is None:
+            raise DGLError('SRC node type "{}" does not exist.'.format(ntype))
+        return ntid
```
Comment (BarclayII) [inline python/dgl/heterograph.py]: Because we sort the node types in `srctypes`, `get_srctype_id(ntype)` may not necessarily return the index of `ntype` in `srctypes`, right?  Would it cause problems?

File: python/dgl/heterograph.py [hunk 3]
```diff
@@ -479,16 +530,75 @@ def get_ntype_id(self, ntype):
...
+        """
+        if ntype is None:
+            if len(self._dsttypes_invmap) != 1:
+                raise DGLError('DST node type name must be specified if there are more than one '
+                               'DST node types.')
+            return 0
+        ntid = self._dsttypes_invmap.get(ntype, None)
+        if ntid is None:
+            raise DGLError('DST node type "{}" does not exist.'.format(ntype))
+        return ntid
```
Comment (BarclayII) [inline python/dgl/heterograph.py]: Same.

File: python/dgl/heterograph.py [hunk 4]
```diff
@@ -3864,6 +4042,55 @@ def make_canonical_etypes(etypes, ntypes, metagraph):
...
+        not uni-bipartite.
+    """
+    src, dst, _ = metagraph.edges()
+    if set(src.tonumpy()).isdisjoint(set(dst.tonumpy())):
+        srctypes = {ntypes[tid] : tid for tid in src}
+        dsttypes = {ntypes[tid] : tid for tid in dst}
+        # handle isolated node types
+        for ntid, ntype in enumerate(ntypes):
+            if ntype not in srctypes and ntype not in dsttypes:
+                srctypes[ntype] = ntid
```
Comment (BarclayII) [inline python/dgl/heterograph.py]: Add this behavior to docstring?

File: python/dgl/transform.py [hunk 5]
```diff
@@ -743,38 +743,46 @@ def compact_graphs(graphs, always_preserve=None):
...
-    Otherwise, the right hand side would be determined by DGL via the rules above.
+    Since DST nodes are included in SRC nodes, a common requirement is to fetch
+    the DST node features from the SRC nodes features. To avoid expensive sparse lookup,
+    the function assures that the DST nodes in both SRC and DST sets have the same ids.
+    As a result, given the node feature tensor ``X`` of type ``utype``,
+    the following code finds the corresponding DST node features of type ``vtype``:
+
+    .. code::
+
+        X[:block.number_of_nodes('vtype')]
```
Comment (BarclayII) [inline python/dgl/transform.py]: If we are converting a block with `'user'` as `utype` and again `'user'` as `vtype` (i.e. I'm converting a homograph to a block), would this statement work?

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+                 num_bases,
+                 num_hidden_layers=1,
+                 dropout=0,
+                 use_self_loop=False):
+        super(EntityClassify, self).__init__()
+        self.g = g
+        self.h_dim = h_dim
+        self.out_dim = out_dim
+        self.rel_names = list(set(g.etypes))
+        self.rel_names.sort()
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: It is to keep a deterministic order so that offline evaluation could work. IMO, it can be avoided as long as we use ParameterDict rather than ParameterList. @classicsong what do you think?

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    if use_cuda:
+        model.cuda()
+
+    # train sampler
+    sampler = HeteroNeighborSampler(g, category, [args.fanout] * args.n_layers)
+    loader = DataLoader(dataset=list(train_idx.numpy()),
+                        batch_size=args.batch_size,
+                        collate_fn=sampler.sample_blocks,
+                        shuffle=True,
+                        num_workers=0)
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: agree

File: python/dgl/heterograph.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/heterograph.py]: Unlike `ntypes`,  `srctypes` and `dsttypes` cannot be used as a mapping from type id to type name. Thus, their results are technically set, but I still return list to keep it similar to `ntypes`. The list is sorted to make the behavior deterministic. Since the type id is retrieved from `_srctypes_invmap`, it won't cause trouble.

File: python/dgl/transform.py (hunk 5 above)
Comment (jermainewang) [reply python/dgl/transform.py]: It should be `X[:block.number_of_nodes('DST/vtype')]`. Will change that.

File: python/dgl/heterograph.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/heterograph.py]: OK.  I think users need to be aware of that though, since `get_srctype_id` would behave very differently from `get_ntype_id`.  Could we make them consistent?

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: I see.  I think using `ParameterDict` can make the code more intuitive.

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+        g.multi_update_all(funcs, 'sum')
+
+        hs = {}
+        for ntype in g.dsttypes:
+            if 'h' in g.dstnodes[ntype].data:
+                hs[ntype] = g.dstnodes[ntype].data['h']
+        def _apply(h):
+            # apply bias and activation
+            if self.self_loop:
+                h = h + th.matmul(xs[ntype], self.loop_weight)
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: You are right. They could have different number of rows.

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+        for layer, block in zip(self.layers, blocks):
+            h = layer(block, h)
+        return h
+
+class HeteroNeighborSampler:
+    def __init__(self, g, category, fanouts, full_neighbor=False):
+        self.g = g
+        self.category = category
+        self.fanouts = fanouts
+        self.full_neighbor = full_neighbor
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: Done

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+
+    def sample_blocks(self, seeds):
+        blocks = []
+        seeds = {self.category : th.tensor(seeds)}
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: ddone

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py]: Done

**Comment by caowencai** [comment]
#3115
